### DISCUSSION TITLE
Is there a way to call cut, copy, paste, and select all from code?

### DISCUSSION THREAD

**Discussion by rthomasv3** [ueberdosis/tiptap#802] [General Questions] [answered]
I'm working on a project that needs a custom context menu for some app specific features, but I'm struggling to find a way to implement the basics of the context menu - like cut, copy, paste, and select all.

Are there commands available for these? Like `editor.commands.copy();` or maybe a way to just send the keystroke `ctrl+c` to the editor? I did see `selectAll` under prosemirror-commands in tiptap-commands package, but I'm not sure how to call that from the editor instance...

If there's not a built-in way to do this, any guidance on a custom extension would be really helpful!

**Comment by hanspagel** [comment, marked as answer]
It’s a ton of work to rebuild the already existing native dropdown. 😬 That said, you can try this:

1. Build a custom Extension, see an example with multiple commands here: https://github.com/ueberdosis/tiptap/blob/main/packages/tiptap-extensions/src/extensions/History.js
2. Import or write commands you need, e. g. `import { selectAll } from prosemirror-commands` and add them to the `commands()` object.
3. Register your extension with tiptap under `extensions`, e. g. `new Editor({extensions: [new MyCustomCommands()]})`
4. Build your dropdown and call `editor.commands.selectAll` and others from there.

Hope that helps!

**Reply by rthomasv3** [reply]
Thanks for the info! And yeah, it's a lot of work, but I think my users will really appreciate the extra features in the end - so it's worth it to me.

I'm going to experiment with making a basic extension that just adds `cut, copy, paste, and selectAll` to the list of commands. I think `replaceText` and `insertText`  from `tiptap-commands` should make `cut` and `paste` possible. Combining that with the clipboard api should all work: https://www.w3.org/TR/clipboard-apis/#async-clipboard-api

I'll post the extension code here when I get something working - maybe others will find it helpful.

**Comment by rthomasv3** [comment]
I was able to get most of the features I need working via `document.execCommand`. It turns out paste and reading the clipboard is completely blocked for security reasons, but I can just show a dialog like Google Docs does saying to use ctrl+v.

The extension itself was as simple as:

```
import { Extension } from 'tiptap'
import { selectAll } from 'prosemirror-commands'

export default class Clipboard extends Extension {

   get name() {
      return 'clipboard'
   }

   commands() {
      return {
         cut: () => this.cut(),
         copy: () => this.copy(),
         selectAll: () => selectAll
      }
   }

   cut() {
      return (state, dispatch) => {
         document.execCommand('cut')
         return true
      }
   }

   copy() {
      return (state, dispatch) => {
         document.execCommand('copy')
         return true
      }
   }
}
```

Then in my vue component on the editor I imported and added `new Clipboard()` to the list of extentions. I also block the default context menu for the `editor-content` element.

```
<editor-content class="editorContent" :editor="editor" spellcheck="false" oncontextmenu="return false;" />
```

Note: `@contextmenu.prevent="handler"` refused to work... not sure why, so I just used `oncontextmenu`.

The menu:
```
<div ref="editorMenu" class="context-menu" v-on-clickaway="hideEditorMenu">
   ...
</div>
```

And for actually showing the context menu (again, `@contextmenu` would have been better...):

```
editorProps: {
	handleClick: (view, pos, event) => {
		if (event.button === 2) {
			   // code for setting up the menu data here (spell suggestions, etc)...

                           // show the menu
			   this.$refs.editorMenu.style.opacity = 1;
			   this.$refs.editorMenu.style.display = 'block';
			   this.$refs.editorMenu.style.left = `${event.pageX}px`;
			   this.$refs.editorMenu.style.top = `${event.pageY}px`;
			});
		}
	}
}
```

I call the commands the same as any others when the context menu buttons are clicked: `this.editor.commands.copy();` etc.

And then the result:

![image](https://user-images.githubusercontent.com/39268276/90550122-20623c00-e155-11ea-8555-0e5652a159e7.png)

The spellchecking extension is a work in progress, but it's mostly done. I based it on Search from `tiptap-extensions` but with an `error` decoration class that underlines instead of a `find` decoration. The actual checking/suggestions is just `nspell`.

Obviously lots of code was left out here, but I'm happy to share more - if anyone has any questions just let me know!

**Reply by hanspagel** [reply]
Thanks for sharing! Amazing work!

**Reply by jlarmstrongiv** [reply]
Wonderful!  Thank you very much for the example

I have updated the extension for Tiptap v2.  Please not that Tiptap v2 now supports [`commands.selectAll();`](https://tiptap.dev/api/commands/select-all#usage) by default.

```ts
import { Extension } from "@tiptap/core";

declare module "@tiptap/core" {
  interface Commands<ReturnType> {
    clipBoard: {
      /**
       * cut selection
       */
      cut: () => ReturnType;
      /**
       * copy selection
       */
      copy: () => ReturnType;
    };
  }
}

const ClipBoard = Extension.create({
  name: "clipBoard",

  // solution https://github.com/ueberdosis/tiptap/discussions/802
  // document.execCommand is deprecated https://github.com/ueberdosis/tiptap/discussions/3129
  // while document.execCommand is deprecated, there is no viable replacement https://stackoverflow.com/a/70831583
  // the new async clipboard API does not support cut, only read and write https://developer.mozilla.org/en-US/docs/Web/API/Clipboard

  addCommands() {
    return {
      cut: () => () => {
        document.execCommand("cut");
        return true;
      },
      copy: () => () => {
        document.execCommand("copy");
        return true;
      },
    };
  },
});
```
